### PR TITLE
Update Scrollbar to support infinite scrollable widget

### LINKED ISSUE #41434: [web]: No scrollbar in (infinite) list widget.
While following and completing the instructions on [Write your first app](https://flutter.dev/docs/get-started/codelab#step-4-create-an-infinite-scrolling-listview), the resulting application does have an infinite scrolling list, but there is no (updating) scroll bar to show the "progress" of the list.

I would expect a scroll bar allowing to navigate from the first element to the "last element in memory", and updating as new content keeps loading.

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes https://github.com/flutter/flutter/issues/41434

CC @Piinks Looking forward to your thoughts.

### Test Demo
```dart
// Copyright 2018 The Flutter team. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:english_words/english_words.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyScrollBehavior extends ScrollBehavior {
  @override
  Widget buildScrollbar(
      BuildContext context, Widget child, ScrollableDetails details) {
    return child;
  }
}

// #docregion MyApp
class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // #docregion build
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      scrollBehavior: MyScrollBehavior(),
      title: 'Startup Name Generator',
      home: RandomWords(),
    );
  }
// #enddocregion build
}
// #enddocregion MyApp

// #docregion RWS-var
class _RandomWordsState extends State<RandomWords> {
  final _suggestions = <WordPair>[];
  final _biggerFont = const TextStyle(fontSize: 18.0);
  final controller = ScrollController();
  // #enddocregion RWS-var

  // #docregion _buildSuggestions
  Widget _buildSuggestions() {
    return ListView.builder(
      // itemCount: 1000,
      key: Key('value'),
      // cacheExtent: 25000,
      // reverse: true,
        controller: controller,
        padding: const EdgeInsets.all(1.0),
        itemBuilder: /*1*/ (context, i) {
          if (i.isOdd) return const Divider(); /*2*/

          final index = i ~/ 2; /*3*/
          if (index >= _suggestions.length) {
            _suggestions.addAll(generateWordPairs().take(1000)); /*4*/
          }
          return _buildRow(_suggestions[index], index);
        });
  }
  // #enddocregion _buildSuggestions

  void _loadData() {
    Future.delayed(const Duration(milliseconds: 200)).then((value) {
      _suggestions.addAll(generateWordPairs().take(10)); /*4*/
    });
  }

  // #docregion _buildRow
  Widget _buildRow(WordPair pair, int index) {
    return ListTile(
      title: Text(
        '$index - ' + pair.asPascalCase,
        style: _biggerFont,
      ),
    );
  }
  // #enddocregion _buildRow

  // #docregion RWS-build
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          controller.jumpTo(700);
        },
      ),
      appBar: AppBar(
        title: const Text('Startup Name Generator'),
      ),
      body: Scrollbar(
        controller: controller,
        thickness: 17,
        isAlwaysShown: true,
        interactive: true,
        child: _buildSuggestions(),
      ),
    );
  }
// #enddocregion RWS-build
// #docregion RWS-var
}
// #enddocregion RWS-var

class RandomWords extends StatefulWidget {
  const RandomWords({Key? key}) : super(key: key);

  @override
  State<RandomWords> createState() => _RandomWordsState();
}
```

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 1]
```diff
@@ -567,11 +568,17 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
   // The size of the thumb track.
   double get _trackExtent => _lastMetrics!.viewportDimension - 2 * mainAxisMargin - _mainAxisPadding;
 
+  double get _cachedContentExtent => math.max(RenderAbstractViewport.defaultCacheExtent, _lastMetrics!.viewportDimension * 2.0);
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scrollbar.dart:571]: Or make this programmable?

**Comment by Piinks** [comment]
Hey @xu-baolin! I have been working on this too, I dusted off a proposal I had for it today, see https://github.com/flutter/flutter/pull/96825
I am cleaning up and old design doc to discuss further too. Maybe we can collaborate with your ideas here for the finished product. :) 
Let me know what you think. I'll add the design doc to the PR once it is ready

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 2]
```diff
@@ -567,11 +568,17 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
...
+
   // The total size of the scrollable content.
   double get _totalContentExtent {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollbar.dart:574]: Ah this is much cleaner than in my proposal.

**Comment by Piinks** [comment]
Hey @xu-baolin I think we're going to move forward with #96825, although I have to push a bunch of changes. I am going to close this, I really appreciate your contributions on this. :)

**Pull request closed without merging**